### PR TITLE
Bvs/standalone mode

### PR DESCRIPTION
**Pull request by bschweiz**
### Resolves

_What Github issue does this resolve (please include link)?_

https://github.com/orgs/ComputeCycles/projects/1#card-61988618

https://github.com/orgs/ComputeCycles/projects/1#card-62046013

### Proposed Changes

_Describe what this Pull Request does_

1. checks on src/index.js to see if the process is from the browser, and if not it initializes a headless scratch project, loads in a default game, and connects to the MQTT bases on the address passes in 

2. to run in standalone mode from the terminal, go to `scratch-vm` in the terminal and run the following command, where (address) should be the IP address for the given MQTT network:     `node src/index.js (address)`

### Reason for Changes

_Explain why these changes should be made_

game can now run and deploy to satellites without the GUI

### Test Coverage

_Please show how you have added tests to cover your changes_
 Tested on local setup

### CODE REVIEW

File: src/virtual-machine.js [hunk 1]
```diff
@@ -345,11 +351,21 @@ class VirtualMachine extends EventEmitter {
         });
         
         this.runtime.on('MQTT_ALIAS_VAR_INBOUND', data => {
-            this.setAliasVariables(data);
+            if (this.standaloneMode === true) {
+                this.setStandaloneAliasVariables(data);
+            }
+            if (this.standaloneMode === false) {
+                this.setAliasVariables(data);
+            }
```
Comment (rvsrvs) [inline src/virtual-machine.js]: if (this.standaloneMode) { } else { }

File: src/virtual-machine.js [hunk 2]
```diff
@@ -345,11 +351,21 @@ class VirtualMachine extends EventEmitter {
...
         });
         
         this.runtime.on('MQTT_GROUP_VAR_INBOUND', data => {
-            this.setGroupVariables(data);
+            if (this.standaloneMode === true) {
+                this.setStandaloneGroupVariables(data);
+            }
+            if (this.standaloneMode === false) {
+                this.setGroupVariables(data);
+            }
```
Comment (rvsrvs) [inline src/virtual-machine.js]: same as above

File: src/virtual-machine.js [hunk 3]
```diff
@@ -568,6 +584,28 @@ class VirtualMachine extends EventEmitter {
...
+            const stage = this.runtime.getTargetForStage();
+            // params (id, name, type, isCloud)
+            stage.createVariable(`${data.alias}`, `${data.alias}`, '', false);
+            setTimeout(() => {
+                stage.variables[data.alias].value = data.payload;
+            }, 100);
+        }
+    }
+
+    setStandaloneGroupVariables (data) {
```
Comment (rvsrvs) [inline src/virtual-machine.js]: I think that what you want to do is set data to be in a standard format above and only have one function that does this stuff

File: src/virtual-machine.js (hunk 1 above)
Comment (rvsrvs) [reply src/virtual-machine.js]: You probably want to get `data` into a standard form before `emit`ting on  `MQTT_ALIAS_VAR_INBOUND` and then I think you don't need the two funcs or the standalone mode flag..  Might still need the flag, but I'd like to see about avoiding it..

File: src/index.js [hunk 4]
```diff
@@ -1,3 +1,35 @@
...
+    virtualMachine.connectMqtt('playspot', `${process.argv[2]}`, '', '');
+    reader.onerror = error => {
+        log.info('No DefaultGame.sb3, exiting');
+        process.exit();
+    };
+    
+    reader.readAsArrayBuffer(file);
+}
+
+virtualMachine.start();
```
Comment (rvsrvs) [inline src/index.js:48]: need to use `yarg` and parse the command line options

**Review by rvsrvs — COMMENTED** [review]
great that we got it working, I think we can get it to the point where we are minimizing our intrusion into scratch-vm.  that way we have fewer changes to maintain

File: src/virtual-machine.js (hunk 3 above)
Comment (bschweiz) [reply src/virtual-machine.js]: that won't work, the .createVariable method that is called without the `workspace` from the GUI is a different method with different params

File: src/virtual-machine.js (hunk 1 above)
Comment (bschweiz) [reply src/virtual-machine.js:359]: if {} else {} syntax actually goes against the Scratch linter controls

File: src/virtual-machine.js (hunk 2 above)
Comment (bschweiz) [reply src/virtual-machine.js]: standaloneMode removed, more simple solution applied

**Pull request merged**